carma_guidance_plugins/include/carma_guidance_plugins/plugin_base_node.hpp

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2022 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#pragma once
+
+#include <gtest/gtest_prod.h>
+#include <rclcpp/rclcpp.hpp>
+#include <carma_planning_msgs/msg/plugin.hpp>
+#include <carma_wm_ros2/WMListener.hpp>
+#include <carma_wm_ros2/WorldModel.hpp>
+
+#include <carma_ros2_utils/carma_lifecycle_node.hpp>
+
+namespace carma_guidance_plugins
+{
+
+  /**
+   * \brief PluginBaseNode provides default functionality for all carma guidance plugins.
+   *        This includes basic state machine management (largely delegated to ROS2 lifecycle behavior), required interfaces, and plugin discovery
+   * 
+   * Extending classes must implement the on_configure_plugin method to load parameters, and my override the other state transitions methods on_<state>_plugin if desired.
+   * Additionally, extending classes must implement the methods such as get_name() which are used to populate the plugin discovery message.
+   * 
+   */
+  class PluginBaseNode : public carma_ros2_utils::CarmaLifecycleNode
+  {
+
+  private:
+    // Publishers
+    //! Publisher for plugin discovery. This publisher will begin publishing 
+    //immediately on node construction regardless of lifecycle state. This is meant to allow for fast plugin activation on startup.
+    rclcpp::Publisher<carma_planning_msgs::msg::Plugin>::SharedPtr plugin_discovery_pub_;
+
+    // Timers
+    //! Timer to trigger publication of the plugin discovery message at a fixed frequency 
+    rclcpp::TimerBase::SharedPtr discovert_timer_;
+
+    // WorldModel listener
+    carma_wm::WMListener wm_listener_;
+
+    // World Model populated by the listener at runtime
+    carma_wm::WorldModelConstPtr wm_;
+
+    /**
+     * \brief Callback for the plugin discovery timer which will publish the plugin discovery message
+     */
+    void discovery_timer_callback();
+
+  public:
+    /**
+     * \brief PluginBaseNode constructor 
+     */
+    explicit PluginBaseNode(const rclcpp::NodeOptions &);
+
+    //! Virtual destructor for safe deletion
+    virtual ~PluginBaseNode() = default;
+
+    /**
+     * \brief Returns the activation status of this plugin.
+     *        The plugins API callbacks will only be triggered when this method returns true. 
+     * 
+     * \return True if plugin is in the ACTIVE state. False otherwise.
+     */ 
+    virtual bool get_activation_status() final;
```

Comment:
Not sure I understand the usage of `virtual ... final` here, they seem at odds. Maybe the implementation will clarify.

Reply:
Yeah this does not need to be virtual.